s3/src/main/scala/akka/stream/alpakka/s3/impl/auth/Signer.scala

```diff
@@ -25,7 +25,9 @@ import akka.stream.scaladsl.Source
         val headersToAdd = Vector(RawHeader("x-amz-date", key.requestDate.format(dateFormatter)),
                                   RawHeader("x-amz-content-sha256", hb)) ++ sessionHeader(key)
         val reqWithHeaders = request.withHeaders(request.headers ++ headersToAdd)
-        val cr = CanonicalRequest.from(reqWithHeaders)
+        val cr = CanonicalRequest.from(
+          reqWithHeaders.withHeaders(reqWithHeaders.headers.filterNot(_.isInstanceOf[`Raw-Request-URI`]))
```

Comment:
It would be more expected to do the filtering directly on the headers in line 27.

Reply:
that's the trick here, the request to build a signature differs from the original request, so we can skip the raw-request-uri synthetic header.

Reply:
should this go to `CanonicalRequest` itself? `CanonicalRequest.from` could filter out known synthetic headers before calling `*HeaderString` methods.

Reply:
I think you can filter out un-rendered headers inside of `CanonicalRequest`:

```scala
  def from(request: HttpRequest): CanonicalRequest = {
    val hashedBody =
      request.headers
        .collectFirst { case header if header.is("x-amz-content-sha256") => header.value }
        .getOrElse("")

    val renderedHeaders = request.headers.filterNot(_.is(`Raw-Request-URI`.lowercaseName))

    CanonicalRequest(
      request.method.value,
      pathEncode(request.uri.path),
      canonicalQueryString(request.uri.query()),
      canonicalHeaderString(renderedHeaders),
      signedHeadersString(renderedHeaders),
      hashedBody
    )
  }
```
Then you could have a spec like this one:

```scala
  it should "correctly build a canonicalString when `Raw-Request-URI` header is present" in {
    val req = HttpRequest(
      HttpMethods.GET,
      Uri("https://s3-eu-central-1.amazonaws.com/my.test.bucket/file+name.txt")
    ).withHeaders(
      RawHeader("x-amz-content-sha256", "testhash"),
      `Content-Type`(ContentTypes.`application/json`),
      `Raw-Request-URI`("/my.test.bucket/file%2Bname.txt")
    )
    val canonical = CanonicalRequest.from(req)
    canonical.canonicalString should equal(
      """GET
        |/my.test.bucket/file%2Bname.txt
        |
        |content-type:application/json
        |x-amz-content-sha256:testhash
        |
        |content-type;x-amz-content-sha256
        |testhash""".stripMargin
    )
  }
```